File: javascript/balanced_brackets.js
function balancedBrackets(string) {
    // store openers and closers in sets for fast access
    const openers = new Set(["(", "[", "{"]);
    const closers = new Set([")", "]", "}"]);
    const pairs = {"(":")", "{":"}", "[":"]"};
    
    // use a stack for LIFO ordering
    // this allows us to check if the brackets adhere to a palindromic ordering
    const stack = [];
    
    // boolean to keep track of whether we currently have an open pipe or not
    let openPipe = false;
      
    for (let i = 0; i < string.length; i++) {
      const char = string[i];
      if (char === "|") {
        openPipe = openPipe ? "NO" : "YES";
      } else if (openers.has(char)) {
        stack.push(char);
      } else if (closers.has(char)) {
      
        // upon encountering a closer, we have a few things to check 
        // 1. if our stack is empty, that means that there is no opener 
        // associated with this closer 
        // 2. does the top opener's type match our closer?
        // 3. is there currently an open pipe?
        if (!stack.length || pairs[stack.pop()] !== char || openPipe) {
          return "NO";
        }
      }
    }
    let retVal = ""
    // after we've walked throughout the entire string, we want to check
    // that our stack is empty and that there are no open pipes
    if (stack.length === 0 && !openPipe) {
        retVal = "YES"
    } else {
        retVal = "NO"
    }
    return retVal
  }

File: python/balanced_brackets.py
def balancedBrackets(string):
    brackets_dict = {'{': '}', '[': ']', '(':')', '|':'|'}
    open_brackets =['{', '[', '(']
    brackets = []
    stack = []

    # Extract all the brackets from the string
    for char in string:
        if char in brackets_dict or char in brackets_dict.values():
            brackets.append(char)
    # Loop through the brackets
    for i, c in enumerate(brackets):
        # Check for consecutive pipes (|)
        # if the pipe isn't in the stack, push it onto the stack
        if c == '|' and '|' not in stack:
            stack.append(c)
        # else, check if the the char is an open brackets
        elif c in open_brackets:
            # push it onto the stack
            stack.append(c)
        # If it's a closing bracket
        else:
            # pop the last open bracket off the stack
            last = stack.pop()
            # Check if the value of the last open bracket in the brackets_dict is equal to the closing bracket
            # Example if c = }, if the last item in the stack is {, then
            # brackets_dict['{'] should be equal to c = }
            if c != brackets_dict[last]:
                # The brackets are not balanced
                return False
    # If the stack is not empty, then the brackets are not balanced
    if len(stack) != 0:
        return False
    # The brackets are balanced
    return True


if __name__ == '__main__':
    print(balancedBrackets('{{||[||]||}}'))